src/main/java/com/mongodb/spark/sql/connector/write/MongoDataWriter.java

```diff
@@ -82,12 +182,11 @@ public void write(final T record) throws IOException {}
    *
    * <p>If this method fails (by throwing an exception), {@link #abort()} will be called and this
    * data writer is considered to have been failed.
-   *
-   * @throws IOException if failure happens during disk/network IO like writing files.
    */
   @Override
-  public WriterCommitMessage commit() throws IOException {
-    return null;
+  public WriterCommitMessage commit() {
```

Comment:
We duplicated `commit`/`abort` documentation on `MongoDataWriter` and `MongoBatchWrite`, as well as duplicated the documentation of these classes from `DataWriter` and `BatchWrite` respectively. However, our `abort` and `commit` do nothing at all related to their intended functionality, thus, as far as my understanding goes, violating at least the requirements for the `DataWriter.commit` method: _"The written data should only be visible to data source readers after BatchWrite.commit(WriterCommitMessage[]) succeeds, which means this method should still "hide" the written data and ask the BatchWrite at driver side to do the final commit via WriterCommitMessage"_.

At the very least we need to fix the documentation in our implementations, which is mostly incorrect because it duplicates the documentation from parent classes/interfaces. Even better, we should make it clear in the user-facing mongo-spark documentation that our implementation of the Spark API does not conform to all the requirements, and specify what requirements are violated.

It is also useful to understand why this requirements exist and what are the potential consequences of violating them (@rozza, maybe you understand this, then it would be useful to have a call). To me the requirements seem weird (maybe because of having no knowledge about Spark):
1. The `DataWriter.commit`/`abort` on each partition (I don't actually know what this is) and the `BatchWrite.commit`/`abort` across partitions seems like an attempt to have something similar to 2PC, but likely broken. 2PC becomes non-trivial when it comes to handling failures, and ["Concurrency Control and Recovery in Database Systems"](https://www.amazon.ca/Concurrency-Control-Recovery-Database-Systems/dp/0201107155) is the only place I bumped into (not saying there are no others) that does a good job explaining the algorithm (the relevant part of the book is shared [here](https://drive.google.com/file/d/1LCThU_bRo_H_jBuN7HHQqMElG9tgHYNG/view?usp=sharing)). Anyway, what happens if an implementation, for example ours, violates the isolation requirement specified by `DataWriter.commit`? It is not clear to me.
2. I cannot make sense of the "speculative tasks" described in the `DataWriter` interface. It seems that the documentation states that a writing task to any partition may be started again, thus running concurrently with another task that writes the same data to the same "place". I can't see how can this generally be possible and what problem is being solved by this. Many DBMS (and certainly all relational DBMS) will just block one of such tasks, and if the tasks are not careful with the order of locks they acquire (usually implicitly, which is what makes it especially dangerous), they may end up deadlocking each other.

Reply:
OK - this comes down to the abilities of different writers.

1. One of the new features of this API is to allow for the possibility of a distributed commit - which wasn't available in the previous API as there weren't the coordination steps. 
  So now there is the ability for each executor (worker) machine can write its data and then the spark driver can coordinate the commit.
  There is a ticket to allow for "hidden" operations by fully supporting a distributed write across the workers and then in the final step renaming the temp table.  However, that would only support a insert only work load and will likely come after release.

2. "speculative tasks" allow Spark to utilise all the executors and first to finish wins. The requirement for this is to idempotent operations - where there's no harm for an operation to be tried again and again.  Deadlocking the DBMS isn't an issue here, this is Spark trying to utilise all its compute and get results as quickly as possible. 
However, our BatchWriter uses the commit coordinator to mitigate against this.

Reply:
**1**
> 1. There is a ticket to allow for "hidden" operations by fully supporting a distributed write across the workers and then in the final step renaming the temp table. However, that would only support a insert only work load and will likely come after release.

This is not documented by the `MongoBatchWrite`. Is there a documentation ticket to document this violation of the spec in mongo-spark docs, and explain what it means for the user? I am asking, because we clearly violate the requirement of the API that we are implementing, and this (as well as consequences) is something that a user has to be aware of.

**2**
> Deadlocking the DBMS isn't an issue here, this is Spark trying to utilise all its compute and get results as quickly as possible.

The amount of computing resources involved does not affect the possibility of having a deadlock (it is either possible or not), it may, if anything, only affect the probability of such an event. Idempotence per se also does not have avoiding deadlocks. `BatchWrite.commit` states _"Spark uses the commit coordinator to allow at most one task to commit"_, but this does not help, because it is both theoretically possible (and, in fact, happens) for a DBMS to acquire implicit locks while executing requests and not only when commit is requested.

> The requirement for this is to idempotent operations

`BatchWrite` does not mention idempotence, where can I see this requirement? My initial understanding was that Spark tries to guarantee that not more than one task of all speculative tasks commits: _"Spark uses the commit coordinator to allow at most one task to commit"_ (from `BatchWrite`). This seems to imply that there is no need for a task to be idempotent.

I am still trying to figure out whether I am misinterpreting something. Could you confirm that the phrase _"speculative execution may cause multiple tasks to run for a partition"_ (from `BatchWrite`) means that the same task may be run concurrently with itself for the same subset of data located in the same place? If yes, then such tasks blocking each other is a possibility in general (may not be a possibility with MongoDB, I know nothing about implicit MongoDB locks to reason about it). If no, then could you please tell what the correct interpretation of the phrase is?

It is also my understanding that we are only "teaching" Spark how to write (this PR) and read (a future PR) to/from MongoDB, and we can't possibly know what tasks will Spark be running and what data/when will be written/read. Is this correct? If this is correct, then we may stop caring about the possibility of tasks deadlocking, because we are not in control of the tasks.

Reply:
1. Its not documented as it has not been implemented. Documentation for the new connector is still in the planning phase.
2. Deadlocks are not a concern here. Its not possible because MongoDB allows multiple processes to write and read from the database at the same time.

Batch write doesnt not mention idempotent operations as a requirement.  It is an implicit requirement (for database operations at least) due to the possibility of speculative execution.

> I am still trying to figure out whether I am misinterpreting something. Could you confirm that the phrase "speculative execution may cause multiple tasks to run for a partition" (from BatchWrite) means that the same task may be run concurrently with itself for the same subset of data located in the same place? If yes, then such tasks blocking each other is a possibility in general (may not be a possibility with MongoDB, I know nothing about implicit MongoDB locks to reason about it). If no, then could you please tell what the correct interpretation of the phrase is?

Spark is a framework that uses multiple machines (executors) to compute data for a given purpose.  The Spark driver is the coordinator of the tasks and sends work to the executors. Spark is also agnostic to connectors - it may be processing data in memory, in files, in a database.

Speculative execution means two separate (executor) machines may get the same (workload) data for a task and will process it independently of each other. This results in duplicated work, the Spark driver will only do this if there is spare compute (executor machines that are idle) and the original executor has taken too long to produce a result eg. there may be a problem with that machine, the network or what ever.   Deadlocking isn't an issue as MongoDB is designed for concurrent writes and reads on data.

> It is also my understanding that we are only "teaching" Spark how to write (this PR) and read (a future PR) to/from MongoDB, and we can't possibly know what tasks will Spark be running and what data/when will be written/read.
Correct.

Reply:
@rozza , thank you for the answers. I described the intention behind the {{BatchWrite}}/{{DataWriter}} requirements, the reasons we cannot meet them, and how exactly we violate them in [DOCSP-19782 [SPARK] Support v2 Write API](https://jira.mongodb.org/browse/DOCSP-19782?focusedCommentId=4199290&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-4199290).I believe, I provided enough information for the documentation team to understand the situation and to write a documentation update. Could you please take a look and say if there is something there you consider incorrect or disagree with?

Reply:
Looks good - will review with the docs team in due course.